File: src/service/itemService.ts
import {Item} from "../models/Item";
import {GetItemsResponse, GenericResponse} from 'types';

export async function GetWantedItems(): Promise<GetItemsResponse> {
    return Item.findAll({where: {reserved: false}, order: [['rank', 'ASC'], ['id', 'DESC']]}).then(items => {
        return {success: true, items: items};
    }).catch(err => {
        return {success: false, items: []};
    });
}

export async function GetReservedItems(reserverid: number): Promise<GetItemsResponse> {
    return Item.findAll({where: {reserved: true, reserverid: reserverid}, order: [['rank', 'ASC'], ['id', 'DESC']]}).then(items => {
        return {success: true, items: items};
    }).catch(err => {
        return {success: false, items: []};
    });
}


// export async function WhoReserved(reserverid: number) {
//     // get first/last name
// }

export async function GetAllItems(): Promise<GetItemsResponse> {
    return Item.scope("user").findAll({order: [['rank', 'ASC'], ['id', 'DESC']]}).then(items => {
        return {success: true, items: items};
    }).catch(err => {
        return {success: false, items: []};
    });
}

export async function AddItem(name: string, url: string, rank: number): Promise<GenericResponse> {
    return Item.create({name: name, url: url, rank: rank}).then(item => {
        if (item) {
            return {success: true, message: "Item created."};
        } else {
            return {success: false, message: "Failed to create item."};
        }
    }).catch(err => {
        console.error(err);
        return {success: false, message: "Failed to create item."};
    })
}

export async function DeleteItem(itemid: number): Promise<GenericResponse> {
    return Item.findOne({where: {id: itemid}}).then(async item => {
        if (item) {
            return await item.destroy().then(itemDeleted => {
                return {success: true, message: "Item Deleted."};
            }).catch(err => {
                return {success: false, message: "Failed to delete item."};
            });
        } else {
            return {success: false, message: "Failed to delete item."};
        }
    })
}

export async function ReserveItem(itemid:number, reserverid: number): Promise<GenericResponse> {
    return Item.findOne({where: {id: itemid}}).then(item => {
        if (item) {
            item.reserved = true;
            item.reserverid = reserverid;
            return item.save().then(item => {
                if (item) {
                    return {success: true, message: "Item Reserved."};
                } else {
                    return {success: false, message: 'Failed to reserve the item.'};
                }

            }).catch(err => {
                console.error(err);
                return {success: false, message: 'Failed to reserve the item.'};
            })
        } else {
            return {success: false, message: "Couldn't find item to reserve."};
        }
    }).catch(err => {
        console.error(err);
        return {success: false, message: "Failed to reserve the item."};
    });
}

export async function UnReserveItem(itemid: number, requesterId: number, requesterUserLevel: number): Promise<GenericResponse> {
    return Item.scope("reserver").findOne({where: {id: itemid}}).then(item => {
        if (item) {
            if ((item.reserverid === requesterId) || (requesterUserLevel === 9)) {
                item.reserved = false;
                item.reserverid = null;
                return item.save().then(item => {
                    if (item) {
                        return {success: true, message: "Item UnReserved."};
                    } else {
                        return {success: false, message: 'Failed to UnReserve the item.'};
                    }

                }).catch(err => {
                    console.error(err);
                    return {success: false, message: 'Failed to UnReserve the item.'};
                })
            } else{
                return {success: false, message: 'Failed to UnReserve the item. You aren\'t the reserver or an admin'};
            }
            
        } else {
            return {success: false, message: "Couldn't find item to UnReserve."};
        }
    }).catch(err => {
        console.error(err);
        return {success: false, message: "Failed to UnReserve the item."};
    });
}

export async function EditItemRank(itemid:number, rank: number): Promise<GenericResponse> {
    return Item.findOne({where: {id: itemid}}).then(item => {
        if (item) {
            item.rank = rank;
            return item.save().then(item => {
                if (item) {
                    return {success: true, message: "Item Rank Updated."};
                } else {
                    return {success: false, message: 'Failed to update the Item Rank.'};
                }

            }).catch(err => {
                console.error(err);
                return {success: false, message: 'Failed to update the Item Rank.'};
            })
        } else {
            return {success: false, message: "Couldn't find item to update the rank.."};
        }
    }).catch(err => {
        console.error(err);
        return {success: false, message: "Failed to update the Item Rank."};
    });
}

File: service/itemService.go
package service

import (
	"github.com/sirupsen/logrus"

	"github.com/jatgam/wishlist-api/models"
	"github.com/jatgam/wishlist-api/types"
)

func GetWantedItems(logger *logrus.Entry) (*[]types.Items, error) {
	wantedItems, err := models.GetWantedItems()
	if err != nil {
		logger.Errorf("GetWantedItems: Failed DB Query: %s", err.Error())
		return nil, types.ErrGetWantedItemsDB
	}
	if wantedItems != nil {
		logger.Infof("Got %v wanted items.", len(*wantedItems))
	} else {
		logger.Info("Got 0 wanted items.")
	}

	return itemDBModelToResponse(wantedItems), nil
}

func GetAllItems(logger *logrus.Entry) (*[]types.Items, error) {
	allItems, err := models.GetAllItems()
	if err != nil {
		logger.Errorf("GetAllItems: Failed DB Query: %s", err.Error())
		return nil, types.ErrGetAllItemsDB
	}
	if allItems != nil {
		logger.Infof("Got %v items", len(*allItems))
	} else {
		logger.Info("Got 0 items")
	}
	return itemDBModelToResponse(allItems), nil
}

func GetReservedItems(userID int, logger *logrus.Entry) (*[]types.Items, error) {
	reservedItems, err := models.GetReservedItems(userID)
	if err != nil {
		logger.Errorf("GetReservedItems: Failed DB Query: %s", err.Error())
		return nil, types.ErrGetReservedItemsDB
	}
	if reservedItems != nil {
		logger.Infof("Got %v items", len(*reservedItems))
	} else {
		logger.Info("Got 0 items")
	}
	return itemDBModelToResponse(reservedItems), nil
}

func EditItemRank(itemID int, rank int, logger *logrus.Entry) error {
	item, err := models.FindOneItem(map[string]interface{}{"id": itemID})
	if err != nil {
		logger.Errorf("EditItemRank: Failed DB Query to find item: %s", err.Error())
		return types.ErrEditItem
	}
	if item == nil {
		logger.Error("EditItemRank: Item Doesn't Exist")
		return types.ErrEditItem
	}
	updateErr := models.UpdateItemWithMap(item, map[string]interface{}{"rank": rank})
	if updateErr != nil {
		logger.Errorf("EditItemRank: Failed DB Query to update item: %s", err.Error())
		return types.ErrEditItem
	}
	return nil
}

func ReserveItem(userID, itemID int, logger *logrus.Entry) error {
	item, err := models.FindOneItem(map[string]interface{}{"id": itemID, "reserved": false}, models.ItemReserveScope)
	if err != nil {
		logger.Errorf("ReserveItem: Failed DB Query to find item: %s", err.Error())
		return types.ErrEditItem
	}
	if item == nil {
		logger.Error("ReserveItem: Item Doesn't Exist, or already reserved")
		return types.ErrEditItem
	}
	updateErr := models.UpdateItemWithMap(item, map[string]interface{}{"reserverid": userID, "reserved": true})
	if updateErr != nil {
		logger.Errorf("ReserveItem: Failed DB Query to update item: %s", err.Error())
		return types.ErrEditItem
	}
	return nil
}

func UnReserveItem(userID, itemID int, logger *logrus.Entry) error {
	item, err := models.FindOneItem(map[string]interface{}{"id": itemID}, models.ItemReserveScope)
	if err != nil {
		logger.Errorf("UnReserveItem: Failed DB Query to find item: %s", err.Error())
		return types.ErrEditItem
	}
	if item == nil {
		logger.Error("UnReserveItem: Item Doesn't Exist")
		return types.ErrEditItem
	}
	if item.ReserverID == nil || *item.ReserverID != userID {
		logger.Errorf("UnReserveItem: Failed, not original reserver or admin")
		return types.ErrEditItem
	}
	updateErr := models.UpdateItemWithMap(item, map[string]interface{}{"reserverid": nil, "reserved": false})
	if updateErr != nil {
		logger.Errorf("UnReserveItem: Failed DB Query to update item: %s", err.Error())
		return types.ErrEditItem
	}
	return nil
}

func AddItem(name, url string, rank int, logger *logrus.Entry) error {
	if err := models.AddItem(name, url, rank); err != nil {
		logger.Errorf("Failed to Add Item: %v, Error: %v", name, err.Error())
		return types.ErrAddItemErr
	}
	return nil
}

func DeleteItem(itemID int, logger *logrus.Entry) error {
	item, err := models.FindOneItem(map[string]interface{}{"id": itemID})
	if err != nil {
		logger.Errorf("DeleteItem: Failed DB Query to find item: %s", err.Error())
		return types.ErrDeleteItem
	}
	if item == nil {
		logger.Error("DeleteItem: Item Doesn't Exist")
		return types.ErrDeleteItem
	}

	if deleteErr := models.DeleteItem(item); deleteErr != nil {
		logger.Errorf("Failed to Delete Item: %v", item.Name)
		return types.ErrDeleteItem
	}

	return nil
}

func itemDBModelToResponse(items *[]models.ItemModel) *[]types.Items {
	var resp []types.Items
	for _, item := range *items {
		resp = append(resp, types.Items{Name: item.Name, Rank: item.Rank, Url: item.URL})
	}
	return &resp
}
